flink-runtime/src/main/java/org/apache/flink/runtime/io/network/partition/SortMergeSubpartitionReader.java

```diff
@@ -176,21 +202,28 @@ public void resumeConsumption() {
 
     @Override
     public Throwable getFailureCause() {
-        // we can never throw an error after this was created
-        return null;
+        synchronized (lock) {
+            return failureCause;
+        }
     }
 
     @Override
     public boolean isAvailable(int numCreditsAvailable) {
-        if (numCreditsAvailable > 0) {
-            return !buffersRead.isEmpty();
-        }
+        synchronized (lock) {
+            if (isReleased) {
+                return true;
+            }
 
-        return !buffersRead.isEmpty() && !buffersRead.peek().isBuffer();
+            if (numCreditsAvailable > 0) {
+                return !buffersRead.isEmpty();
+            }
+
+            return !buffersRead.isEmpty() && !buffersRead.peek().isBuffer();
+        }
     }
 
     @Override
     public int unsynchronizedGetNumberOfQueuedBuffers() {
-        return 0;
+        return buffersRead.size();
```

Comment:
why don't we need a lock for it?

Reply:
Because the method has unsynchronized prefix, I think the reason is to avoid getting metrics influence the performance. The PipelinedSubpartition does similar things. Maybe we should also use Math.max(buffersRead.size(), 0) like the PipelinedSubpartition.